datafusion/core/src/physical_plan/union.rs

```diff
@@ -220,55 +220,55 @@ impl ExecutionPlan for UnionExec {
     }
 
     fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
-        let first_input_ordering = self.inputs[0].output_ordering();
-        // If the Union is not partition aware and all the input ordering spec strictly equal with the first_input_ordering
-        // Return the first_input_ordering as the output_ordering
-        //
-        // It might be too strict here in the case that the input ordering are compatible but not exactly the same.
-        // For example one input ordering has the ordering spec SortExpr('a','b','c') and the other has the ordering
-        // spec SortExpr('a'), It is safe to derive the out ordering with the spec SortExpr('a').
-        if !self.partition_aware
-            && first_input_ordering.is_some()
-            && self
-                .inputs
-                .iter()
-                .map(|plan| plan.output_ordering())
-                .all(|ordering| {
-                    ordering.is_some()
-                        && sort_expr_list_eq_strict_order(
-                            ordering.unwrap(),
-                            first_input_ordering.unwrap(),
-                        )
-                })
-        {
-            first_input_ordering
+        // If the Union is partition aware, there is no output ordering.
+        // Otherwise, the output ordering is the "meet" of its input orderings.
+        // The meet is the finest ordering that satisfied by all the input
+        // orderings, see https://en.wikipedia.org/wiki/Join_and_meet.
+        if self.partition_aware {
+            return None;
+        }
+        // To find the meet, we first find the smallest input ordering.
+        let mut smallest: Option<&[PhysicalSortExpr]> = None;
```

Comment:
I separated functionality and made it a function in the file datafusion/core/src/physical_plan/common.rs. Unfortunately at datafusion/physical-expr/src/utils.rs I couldn't import `ExecutionPlan`.

Reply:
Hi @alamb @mustafasrepo 

I think the implementation of the new method `get_meet_of_orderings()` is incorrect. 
And there are two issues:

1)  One child has the sort keys ('a','b','c'), another child has the sort keys ('a','x','y'), this method should return ('a')
2) The usage of `child.equivalence_properties()`is problematic.  The equivalence properties relationship only apply to that specific child plan, we can not inter-apply this relationship among different child plans.

```rust
  // Check if the smallest ordering is a meet or not:
    if children.iter().all(|child| {
        ordering_satisfy(child.output_ordering(), smallest, || {
            child.equivalence_properties()
        })
    })
```

Reply:
Regarding 1, the initial implementation treats input orderings "atomically", so it finds the meet in that sense. I think you are suggesting we find the meet by "looking inside". So the initial implementation is overly conservative; i.e. wrong in a harmless way (at least in the use cases I can think of). But I agree that we should fix it.

Regarding 2, what would be your suggestion? Should we check ordering without any normalization?

Happy to fix this quickly based on your feedback. Thanks for the detailed review, and no worries about the late response.

Reply:
> Regarding 1, the initial implementation treats input orderings "atomically", so it finds the meet in that sense. I think you are suggesting we find the meet by "looking inside". So the initial implementation is overly conservative; i.e. wrong in a relatively harmless way (it will lose fine ordering information but not come up with a wrong ordering). Nevertheless, I agree that we should fix it.
> 
> Regarding 2, what would be your suggestion? Should we check ordering without any normalization?
> 
> Happy to fix this quickly based on your feedback. Thanks for the detailed review, and no worries about the late response.

Regarding 1,  yes, the ideal logic is to find out the common longest prefix among all the sorting keys. not sure whether it is so called meet in math, sorry that I'm not good at math.
Regarding 2, yes, you can just check the ordering without any normalization.

Reply:
@mingmwang, I spent a little bit of time thinking how this can be implemented. Ignoring equivalences, I came up with this:
```rust
let mut idx = 0;
let first = orderings[0];
loop {
    for ordering in orderings.iter() {
        if idx >= ordering.len() {
            return Some(ordering);
        } else if ordering[idx] != first[idx] {
            return if idx > 0 {
                Some(&ordering[..idx])
            } else {
                None
            };
        }
    }
    idx += 1;
}
```
I wrote some tests for this and it seems to work for various cases (including your example). What do you think? 

Circling back to equivalences, I am not sure if such handling is necessary or sensible in this case -- maybe you can provide some insight about that. If such handling is necessary, we will need to replace the not-equality check `ordering[idx] != first[idx]` with an equivalence-aware check. However, at this point, I am not sure what to replace it with (if it is indeed necessary). If you can help out with these points, we can quickly make a PR and get the fix in. Thanks.

Reply:
I think we do not need the equivalence-aware check here.  You implementation LGTM.

Reply:
Thanks for double checking, will open a PR shortly 👍